core/src/main/java/org/apache/iceberg/rest/HttpRESTClient.java

```diff
@@ -0,0 +1,261 @@
+    String fullUri = String.format("%s/%s", uri, path);
+    HttpUriRequestBase request = new HttpUriRequestBase(method.name(), URI.create(fullUri));
+    addRequestHeaders(request);
+
+    if (body != null) {
+      try {
+        StringEntity stringEntity = new StringEntity(mapper.writeValueAsString(body));
+        request.setEntity(stringEntity);
+      } catch (JsonProcessingException e) {
+        throw new RESTException(e, "Failed to write request body: %s", body);
+      }
+    }
+
+    requestInterceptor.accept(request);
+
+    try (CloseableHttpResponse response = httpClient.execute(request)) {
+
+      // HEAD request.
+      if (responseType == null) {
+        if (response.getCode() < 300) {
+          return null;
+        } else {
+          throw new RESTException(
+              "Received a non-2xx response for a request that was not expecting a request body");
+        }
+      }
+
+      if (response.getCode() != HttpStatus.SC_OK && response.getCode() >= 400) {
+        ErrorResponse parsedError;
+        String responseBody = null;
+        try {
+          responseBody = extractResponseBodyAsString(response);
+          parsedError = ErrorResponseParser.fromJson(responseBody);
+          errorHandler.accept(parsedError);
+        } catch (UncheckedIOException e) {
+          throw new RESTException(e,
+              "Received a non-200 response but could not parse a standard error response: %s", response);
+        }
+      }
+
+      HttpEntity entity = response.getEntity();
+      return mapper.readValue(entity.getContent(), responseType);
+    } catch (IOException e) {
+      String desiredType = responseType == null ? "null" : responseType.getSimpleName();
+      throw new UncheckedIOException(
+          String.format("Unhandled exception trying to read in data of type %s", desiredType), e);
+    }
+  }
+
+  @Override
+  public void head(String path, Consumer<ErrorResponse> errorHandler) {
+    execute(Method.HEAD, path, null, null, errorHandler);
+  }
+
+  @Override
+  public <T> T get(String path, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
+    return execute(Method.GET, path, null, responseType, errorHandler);
+  }
+
+  @Override
+  public <T> T post(String path, Object body, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
+    return execute(Method.POST, path, body, responseType, errorHandler);
+  }
+
+  @Override
+  public <T> T delete(String path, Class<T> responseType, Consumer<ErrorResponse> errorHandler) {
+    return execute(Method.DELETE, path, null, responseType, errorHandler);
+  }
+
+  private void addRequestHeaders(HttpUriRequest request) {
+    request.setHeader(HttpHeaders.ACCEPT, ContentType.APPLICATION_JSON.getMimeType());
+    // Many systems require tht content type is set regardless and will fail, even on an empty bodied request.
+    request.setHeader(HttpHeaders.CONTENT_TYPE, ContentType.APPLICATION_JSON.getMimeType());
```

Comment:
I've noticed that some servers will fail if `Content-Type` is not supplied at all (even if there is no body).

I'm wondering if anybody knows of any systems that will fail if `Content-Type` is set and there is no body?